### PR TITLE
Test: withdraw unit tests

### LINKED ISSUE #491: StakingPool unit tests: withdraw
**withdraw**
- [ ] reverts if trying to withdraw stake locked in governance
- [ ] allows to withdraw only stake
- [ ] allows to withdraw only rewards
- [ ] allows to withdraw stake and rewards from multiple tranches
- [ ] updates tranches
- [ ] updates global variables totalWithdrawnStake and totalWithdrawnRewards
- [ ] transfers nxm stake and rewards from token controller to nft owner
- [ ] allows to withdraw stake only if tranche is expired
- [ ] anyone can call to withdraw stake and rewards for a token id
- [ ] Should emit some event

### PR DESCRIPTION
**Pull request by paterson1**
## Context

Closes #491 

## Changes proposed in this pull request

Added `Withdraw` event to `StakingPool` contract and `withdrawNXMStakeAndRewards` method to `TokenControllerMock`. Also in setup update `master` address for tokenController.

## Test plan

Added unit tests to `StakingPool/withdraw.js`.

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: test/unit/StakingPool/withdraw.js [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+
+    await expect(
+      stakingPool.connect(user).withdraw([{ tokenId: depositNftId, withdrawStake, withdrawRewards, trancheIds }]),
+    )
+      .to.emit(stakingPool, 'Withdraw')
+      .withArgs(user.address, depositNftId, trancheIds[1], stakes[1], rewards[1])
+      .to.emit(stakingPool, 'Withdraw')
+      .withArgs(user.address, depositNftId, trancheIds[2], stakes[2], rewards[2])
+      .to.emit(stakingPool, 'Withdraw')
+      .withArgs(user.address, depositNftId, trancheIds[2], stakes[2], rewards[2]);
```
Comment (paterson1) [inline test/unit/StakingPool/withdraw.js:571]: This should have index 0, I'm still debugging why is not the same as the emitted args.

**Comment by patitonar** [comment]
> Pushed a failed test.

While reviewing this test I noticed that the total amount of minted rewards was not being distributed to the users and manager, there was a small amount that remained locked in the contract. After debugging I found that the cause of this small amount was a precision loss in the calculation of `accNxmPerRewardsShare`. 

As it is calculated with the following formula `elapsed * _rewardPerSecond / _rewardsSharesSupply` there were a lot of decimals truncated causing the accumulated value to be less than the actual amount to distribute. The solution I thought for this was to calculate and store the accumulator with an increased precision of ONE_NXM and then scale back the value when calculating the amount of the rewards to transfer.

I applied that change in the commit [Increase accNxmPerRewardsShare precision](https://github.com/NexusMutual/smart-contracts/pull/529/commits/f930209d3ec98d6f45411d28e7aad26238a8d9e1)

Let me know your thoughts on this @shark0der

**Comment by shark0der** [comment]
The `accNxmPerRewardsShare` fix looks good to me. In terms of semantics I think the previous value was:
> the accumulated nxm per **1 reward share**

where "1" was actually "1 wei"

The new value represents
> the accmuluated nxm per **1e18 reward shares**

which makes perfect sense and solves the precision loss issue.

**Review by shark0der — APPROVED** [review]

**Review by shark0der — CHANGES_REQUESTED** [review]
Right now it seems to me that all tests are using a single staker (unless I missed something). Could you please add a new test where 2+ stakers would deposit and withdraw their stake and rewards? I'm not expecting to discover any issues, but just in case, you know!

**Review by roxdanila — APPROVED** [review]

**Review by shark0der — APPROVED** [review]

**Pull request merged**